x-pack/plugins/ml/server/models/calculate_model_memory_limit/calculate_model_memory_limit.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import numeral from '@elastic/numeral';
+import { TypeOf } from '@kbn/config-schema';
+import { APICaller } from 'kibana/server';
+import { analysisConfigSchema } from '../../routes/schemas/anomaly_detectors_schema';
+import { fieldsServiceProvider } from '../fields_service';
+
+interface ModelMemoryEstimationResult {
+  /**
+   * Result model memory limit
+   */
+  modelMemoryLimit: string;
+  /**
+   * Estimated model memory by elasticsearch ml endpoint
+   */
+  estimatedModelMemoryLimit: string;
+  /**
+   * Maximum model memory limit
+   */
+  maxModelMemoryLimit?: string;
+}
+
+/**
+ * Response of the _estimate_model_memory endpoint.
+ */
+export interface ModelMemoryEstimate {
+  model_memory_estimate: string;
+}
+
+export function calculateModelMemoryLimitProvider(callAsCurrentUser: APICaller) {
+  const fieldsService = fieldsServiceProvider(callAsCurrentUser);
+
+  /**
+   * Retrieves an estimated size of the model memory limit used in the job config
+   * based on the cardinality of the fields being used to split the data
+   * and influencers.
+   */
+  return async function calculateModelMemoryLimit(
+    analysisConfig: TypeOf<typeof analysisConfigSchema>,
+    indexPattern: string,
+    query: any,
+    timeFieldName: string,
+    earliestMs: number,
+    latestMs: number,
+    allowMMLGreaterThanMax = false
+  ): Promise<ModelMemoryEstimationResult> {
+    let maxModelMemoryLimit;
+    try {
+      const resp = await callAsCurrentUser('ml.info');
+      if (resp?.limits?.max_model_memory_limit !== undefined) {
+        maxModelMemoryLimit = resp.limits.max_model_memory_limit.toUpperCase();
+      }
+    } catch (e) {
+      throw new Error('Unable to retrieve max model memory limit');
+    }
+
+    let overallCardinality = {};
+    let maxBucketCardinality = {};
+    try {
+      const { influencers, detectors } = analysisConfig;
+
+      const overallCardinalityFields: Set<string> = detectors.reduce(
+        (
+          acc,
+          {
+            by_field_name: baseFieldName,
+            partition_field_name: partitionFieldName,
+            over_field_name: overFieldName,
+          }
+        ) => {
+          [baseFieldName, partitionFieldName, overFieldName]
+            .filter(field => !!field !== undefined && field !== '')
```

Comment:
You should probably filter out `mlcategory` here (unless that's done somewhere else).  `mlcategory` is the magic field name used to mean the output of categorization, so it will have cardinality zero in the actual input data.

Reply:
Fixed in https://github.com/elastic/kibana/pull/60376/commits/c974ae29dc71b9e93a4fe7a62e7aaeae50eb80db